### PR TITLE
Treat <title>, <script>, and <style> contents as text

### LINKED ISSUE #21: Skip tokenization for text in script tags
I understand simple-html-tokenizer is not going to tokenize script tags, but it seems it should not throw an exception on them. I started in htmlbars repo but was directed here. See:

http://emberjs.jsbin.com/jocafazaxa/1/edit?html,js,console

Expected behavior here would be that the sort function is available on the window object.

Note that changing the operator to a '>' works fine.

If simple-html-tokenizer were to simply recognize it is inside of a script tag and to skip tokenization until a closing script tag is found, this would be sufficient.

### PR DESCRIPTION
**Pull request by CvX**
Fixes #21 

This PR is one of the fixes that aim to improve <title>, <script>, and <style> support in Ember FastBoot (see: https://github.com/glimmerjs/glimmer-vm/issues/796#issuecomment-506818798)

cc: @rwjblue

### CODE REVIEW

**Comment by josemarluedke** [comment]
Tests look good! Looking forward to having these fixes in Fastboot.

File: src/evented-tokenizer.ts [hunk 1]
```diff
@@ -113,21 +114,31 @@ export default class EventedTokenizer {
     this.delegate.appendToTagName(char);
   }
 
+  private isIgnoredEndTag(): boolean {
+    let tag = this.tagNameBuffer.toLowerCase();
+
+    return (tag === 'title' && this.input.substr(this.index, 8) !== '</title>') ||
```
Comment (rwjblue) [inline src/evented-tokenizer.ts]: Lets avoid [String.prototype.substr](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) here (since its "deprecated"), maybe use `this.input.substring(this.index, this.index + 8)`?

File: tests/tokenizer-tests.ts [hunk 2]
```diff
@@ -217,6 +222,38 @@ QUnit.test('A newline immediately following a <textarea> tag is stripped', funct
...
+// https://html.spec.whatwg.org/multipage/semantics.html#the-style-element
+QUnit.test('The style element content is always text', function(assert) {
+  let tokens = tokenize("<style>&quot;hey <b>there</b><!-- comment --></style>");
+  assert.deepEqual(tokens, [startTag('style'), chars('"hey <b>there</b><!-- comment -->'), endTag('style')]);
+});
+
+// https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements
+QUnit.test('The script element content restrictions', function(assert) {
+  let tokens = tokenize("<script>&quot;hey <b>there</b><!-- comment --></script>");
+  assert.deepEqual(tokens, [startTag('script'), chars('"hey <b>there</b><!-- comment -->'), endTag('script')]);
```
Comment (rwjblue) [inline tests/tokenizer-tests.ts]: > chars('"hey <b>there</b>')

It is not clear to me that we actually want the `&quot;` to be transformed into `"` (for `script` and `style` that is), I _think_ we should pass it through without transforming the char refs. Have I misread the spec?

File: src/evented-tokenizer.ts (hunk 1 above)
Comment (CvX) [reply src/evented-tokenizer.ts]: Good catch, my bad! Fixed.

File: tests/tokenizer-tests.ts (hunk 2 above)
Comment (CvX) [reply tests/tokenizer-tests.ts]: Ah, the copy&paste TDD! 😅 I took the `<title>` test, and used it as a base for script and style tests, and then the implementation… Working on a fix. 😉

**Comment by CvX** [comment]
Btw. I didn’t implement the `<script>` behavior I’ve linked to in the test (https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements).
IMO it’s an edge case (“for historical reasons (…) is a strange and exotic practice that acts unintuitively”, to quote the spec) that can be ignored.

**Comment by rwjblue** [comment]
Gonna leave this open for a little bit (ping me if you haven't heard back in a day or so) so @krisselden has a chance to review.

**Comment by josemarluedke** [comment]
@rwjblue friendly ping on this one. :D

**Comment by rwjblue** [comment]
Released as https://github.com/tildeio/simple-html-tokenizer/releases/tag/v0.5.8

We'll need to bump the minimum version over in glimmerjs/glimmer-vm, and do a release. Does anyone have time to do that?

**Comment by josemarluedke** [comment]
@rwjblue Got a PR up for that https://github.com/glimmerjs/glimmer-vm/pull/960, note that it ended up having some test failures in the VM.

**Comment by brodybits** [comment]
This looks like a breaking change, see https://github.com/prettier/prettier/pull/6570#issuecomment-550266720

/cc @fisker

**Comment by rwjblue** [comment]
@brodybits - Mind reporting an issue so we can discuss and track the path forward a bit easier?

**Comment by brodybits** [comment]
> Mind reporting an issue so we can discuss and track the path forward a bit easier?

Sure, but there will not be limited detail until I get a chance to investigate it deeper.

Hello from Providence. What cheer netop?

**Comment by rwjblue** [comment]
Hello! 🎵 It's a small world after all... 🎵 😸
